velox/functions/sparksql/SplitFunctions.cpp

```diff
@@ -16,124 +16,226 @@
+#include "velox/functions/lib/Re2Functions.h"
 
 namespace facebook::velox::functions::sparksql {
 namespace {
 
-/// This class only implements the basic split version in which the pattern is a
-/// single character
-class SplitCharacter final : public exec::VectorFunction {
+/**
+ * split(string, delimiter[, limit]) -> array(varchar)
+ *
+ * Splits string on delimiter and returns an array of size at most limit.
+ * delimiter is a string representing regular expression.
+ * limit is an integer which controls the number of times the regex is applied.
+ * By default, limit is -1.
+ *
+ * If delimiter is not empty, then when limit > 0, the last element in
+ * the array will contain the remainder of the string if such left, otherwise
+ * not.
+ *
+ * If limit <= 0, delimiter will be applied as many times as possible, and the
+ resulting array can be of any size.
+ */
+class Split final : public exec::VectorFunction {
  public:
-  explicit SplitCharacter(const char pattern) : pattern_{pattern} {
-    static constexpr std::string_view kRegexChars = ".$|()[{^?*+\\";
-    VELOX_CHECK(
-        kRegexChars.find(pattern) == std::string::npos,
-        "This version of split supports single-length non-regex patterns");
-  }
+  Split() {}
 
   void apply(
       const SelectivityVector& rows,
       std::vector<VectorPtr>& args,
       const TypePtr& /* outputType */,
       exec::EvalCtx& context,
       VectorPtr& result) const override {
-    exec::LocalDecodedVector input(context, *args[0], rows);
+    // Get the decoded vectors out of arguments.
+    exec::DecodedArgs decodedArgs(rows, args, context);
+    DecodedVector* strings = decodedArgs.at(0);
+    DecodedVector* delims = decodedArgs.at(1);
+    DecodedVector* limits = (args.size() == 2) ? nullptr : decodedArgs.at(2);
 
     BaseVector::ensureWritable(rows, ARRAY(VARCHAR()), context.pool(), result);
     exec::VectorWriter<Array<Varchar>> resultWriter;
     resultWriter.init(*result->as<ArrayVector>());
-
-    rows.applyToSelected([&](vector_size_t row) {
-      resultWriter.setOffset(row);
-      auto& arrayWriter = resultWriter.current();
-
-      const StringView& current = input->valueAt<StringView>(row);
-      const char* pos = current.begin();
-      const char* end = pos + current.size();
-      const char* delim;
-      do {
-        delim = std::find(pos, end, pattern_);
-        arrayWriter.add_item().setNoCopy(StringView(pos, delim - pos));
-        pos = delim + 1; // Skip past delim.
-      } while (delim != end);
-
-      resultWriter.commit();
-    });
+    // Optimization for the (flat, const, const) case.
+    if (strings->isIdentityMapping() and delims->isConstantMapping() and
+        (!limits or limits->isConstantMapping())) {
+      const auto* rawStrings = strings->data<StringView>();
+      const auto delim = delims->valueAt<StringView>(0);
+      int32_t limit = std::numeric_limits<int32_t>::max();
+      bool isEmptyDelimiter = delim.empty();
```

Comment:
nit: to improve readability, move this line after L73 because this part is deciding the limit.